lib/IRGen/IRGenSIL.cpp

```diff
@@ -2417,6 +2451,18 @@ void IRGenSILFunction::visitFullApplySite(FullApplySite site) {
     diffCalleeOrigFnLV = LoweredValue(origFnExplosion);
   }
 
+  if (auto fn = site.getInitiallyReferencedFunction()) {
+    auto clangDecl = fn->getClangDecl();
+    if (clangDecl && isa<clang::FunctionTemplateDecl>(clangDecl)) {
+      setCXXFunctionTemplateAddr(
```

Comment:
We ought to aim to do this much, much earlier in the process here. This should be done as part of the lowering of a dedicated SIL instruction.